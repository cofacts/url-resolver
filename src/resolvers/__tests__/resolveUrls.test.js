const { gql } = require('../../../tests/util');
const { getBrowserPromise } = require('../../lib/scrap');

describe('resolveUrls', () => {
  it('resolves normal URLs', async () => {
    const result = await gql`
      {
        resolvedUrls(urls: ["https://example.com"]) {
          url
          canonical
          title
          summary
          html
          topImageUrl
          status
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('returns fetch error properly', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "https://this-cannot-be-resolved.com"
            "line://ch/1341209850"
            "https://identityredesign.tw/vote-list.html" # has domain, but don't respond
            "http://beingsweetlife.com/archives/04/220087" # socket hang up without response (ERR_EMPTY_RESPONSE)
            "https://www.ey.gov.tw/File/66E9E54960EB958B?A=C" # PDF file, which is not supported
            "https://organizejobs.net/en/support.php" # Cert error
          ]
        ) {
          url
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it.skip('resolves alert boxes', async () => {});

  it.skip('resolves 301 redirects', async () => {});

  it('resolves Javascript / html redirects', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "https://www.tstartel.com/static/discount/20161020_rollover/index.htm?utm_source=line&utm_medium=pic&utm_content=line_0504.rollover" # uses window.location.href
            "https://ikeabc.ecrm.com.tw/ikeaoa/l/11166" # uses window.location.replace()
            "https://www.gvm.com.tw/article_content_33230.html" # uses <meta /> refresh
          ]
        ) {
          url
          canonical
          title
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('handles youtube URL', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "https://www.youtube.com/watch?v=jNQXAC9IVRw"
            "https://www.youtube.com/watch?v=not-exist"
          ]
        ) {
          url
          canonical
          title
          summary
          status
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('returns minimal data on error', async () => {
    const result = await gql`
      {
        resolvedUrls(
          urls: [
            "http://blog.udn.com/watercmd/1066441" # https://github.com/cofacts/url-resolver/issues/2
            "https://pension.president.gov.tw/cp.aspx?n=0710ED8C9356A871" # This page overrides URL and causes error when fetching topImageUrl...
            "http://ms7.tw/DL/D?k=app_daily_coupon" # This page don't have error, but executor() returns nothing
            "https://99md.cn/fslh" # Refuses bot connection (connect ECONNREFUSED)
          ]
        ) {
          url
          canonical
          title
          summary
          topImageUrl
          error
        }
      }
    `();

    expect(result).toMatchSnapshot();
  });

  it('handles URLs without protocol', async () => {
    const result = await gql`
      {
        resolvedUrls(urls: ["example.com"]) {
          url
          canonical
          title
        }
      }
    `();

    // Note that url should be same as input.
    //
    expect(result).toMatchSnapshot();
  });
});

afterAll(async () => {
  const browser = await getBrowserPromise();
  await browser.close();

  // eslint-disable-next-line no-console
  console.log('browser closed.');
});
