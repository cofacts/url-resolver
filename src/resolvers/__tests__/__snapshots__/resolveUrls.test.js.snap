// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveUrls handles URLs without protocol 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "http://example.com/",
        "title": "Example Domain",
        "url": "example.com",
      },
    ],
  },
}
`;

exports[`resolveUrls handles youtube URL 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://youtu.be/jNQXAC9IVRw",
        "error": null,
        "status": 200,
        "summary": "The first video on YouTube. Maybe it's time to go back to the zoo? The name of the music playing in the background is Darude - Sandstorm.",
        "title": "Me at the zoo",
        "url": "https://www.youtube.com/watch?v=jNQXAC9IVRw",
      },
      Object {
        "canonical": null,
        "error": null,
        "status": 404,
        "summary": null,
        "title": null,
        "url": "https://www.youtube.com/watch?v=not-exist",
      },
    ],
  },
}
`;

exports[`resolveUrls resolves Javascript / html redirects 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://www.tstartel.com/CWS/allrates-limitedtallk.php",
        "title": "- 台灣之星TSTAR",
        "url": "https://www.tstartel.com/static/discount/20161020_rollover/index.htm?utm_source=line&utm_medium=pic&utm_content=line_0504.rollover",
      },
      Object {
        "canonical": "",
        "title": "宜家卡 - IKEA",
        "url": "https://ikeabc.ecrm.com.tw/ikeaoa/l/11166",
      },
      Object {
        "canonical": "https://www.gvm.com.tw/article.html?id=42308",
        "title": "小心，愈來愈多人跟台灣說再見| 遠見雜誌",
        "url": "https://www.gvm.com.tw/article_content_33230.html",
      },
    ],
  },
}
`;

exports[`resolveUrls resolves normal URLs 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "https://example.com/",
        "error": null,
        "html": "<!DOCTYPE html><html><head>
    <title>Example Domain</title>

    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <style type=\\"text/css\\">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: \\"Open Sans\\", \\"Helvetica Neue\\", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 50px;
        background-color: #fff;
        border-radius: 1em;
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        body {
            background-color: #fff;
        }
        div {
            width: auto;
            margin: 0 auto;
            border-radius: 0;
            padding: 1em;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domain is established to be used for illustrative examples in documents. You may use this
    domain in examples without prior coordination or asking for permission.</p>
    <p><a href=\\"http://www.iana.org/domains/example\\">More information...</a></p>
</div>


</body></html>",
        "status": 200,
        "summary": "This domain is established to be used for illustrative examples in documents. You may use this
    domain in examples without prior coordination or asking for permission.
    More information...",
        "title": "Example Domain",
        "topImageUrl": "",
        "url": "https://example.com",
      },
    ],
  },
}
`;

exports[`resolveUrls returns fetch error properly 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "error": "NAME_NOT_RESOLVED",
        "url": "https://this-cannot-be-resolved.com",
      },
      Object {
        "error": "UNSUPPORTED",
        "url": "line://ch/1341209850",
      },
      Object {
        "error": "NOT_REACHABLE",
        "url": "https://identityredesign.tw/vote-list.html",
      },
      Object {
        "error": "NOT_REACHABLE",
        "url": "http://beingsweetlife.com/archives/04/220087",
      },
      Object {
        "error": "UNSUPPORTED",
        "url": "https://www.ey.gov.tw/File/66E9E54960EB958B?A=C",
      },
      Object {
        "error": "HTTPS_ERROR",
        "url": "https://organizejobs.net/en/support.php",
      },
      Object {
        "error": "NOT_REACHABLE",
        "url": "https://99md.cn/fslh",
      },
    ],
  },
}
`;

exports[`resolveUrls returns minimal data on error 1`] = `
Object {
  "data": Object {
    "resolvedUrls": Array [
      Object {
        "canonical": "http://blog.udn.com/watercmd/1066441",
        "error": null,
        "summary": "轉貼:　　導正『吃香腸配養樂多會致癌嗎？』的錯誤觀念　陳柏瑞　　硝酸鹽和亞硝酸鹽是廣泛存在於自然環境中的化學物質，特別是在食物中，如蔬菜、肉類和魚類都含有一定量的亞硝酸鹽。蔬菜要在生長過程中合成必要的植物蛋白，就要吸收土壤中的硝酸鹽；醃製食品如魚和肉中含有亞硝酸鹽。　　食品加工行業將亞硝酸鹽作為食品添加劑使用有數十年的歷史了，在肉類食品製品中，烤香腸、臘肉、火腿、熱狗、培根等食物含硝酸鹽及亞硝酸鹽殘留量最高不會超過50mg/ Kg。一般醃製10天後亞硝酸鹽殘留量開始下降，食鹽在醃製食品過程中產生抑菌防腐的作用，當濃度10~15％時，只有少數細菌生長，當濃度超過2! 0%時，幾乎所有的微生物都會停止生長。　　膳食中的絕大部分亞硝酸鹽在人體內以“過客”的形式隨尿排出，只有在特定的條件下轉化成為亞硝胺(一種已知的致癌物質)。這種特定的條件包括環境酸鹼度、微生物菌群和適宜的溫度。食物入口後，口中的細菌(硝酸鹽還原菌)會將食物中的硝酸鹽還原為亞硝酸鹽，而亞硝酸鹽在胃腸道的酸性環境中是有機會被轉化為具有致癌作用的亞硝胺，若與富含胺類的海鮮合吃，就會在體內合成大量的亞硝胺。醫界證實，亞硝胺也會阻礙紅血球攜氧功\\\\能，所以人體吸收過量就會感到疲累、體力下降。　　1994年聯合國糧農組織以及世界衛生組織規定硝酸鹽和亞硝酸鹽的每日允許\\\\攝入量（ADI值）分別為 5mg / kg.bw和0.2mg / kg.bw。人體攝入0.3 ~ 0.5g亞硝酸鹽即可引起中毒，3g可致死。不過一般膳食中的亞硝酸鹽是不會對人體健康造成危害，因為造成亞硝酸鹽中毒的劑量是 200毫克以上，而正常膳食中的亞硝酸鹽即使是在食用了大量用亞硝酸鹽做食品添加劑的肉類，也不會超過50毫克。至於膳食中的亞硝酸鹽的“致癌作用”，是人們把亞硝酸鹽與亞硝胺這兩種不同! 的化學物質混淆了，而產生極大的誤解。　　的確，亞硝! 胺的致癌作用是有許\\\\多科學研究做依據的。但是，食品中的亞硝酸鹽含量不等於亞硝胺的含量。食品中的亞硝酸鹽限制量也絕對不同於亞硝胺限制量；而且目前並無亞硝胺對人類致癌的直接證據。世界各地的中毒事件，大多是誤將亞硝酸鹽當作食鹽使用導致過量引起急性中毒的發生。　　因此，只要我們按照平衡膳食的原則，安排我們的日常膳食，就能夠控制住亞硝酸鹽可能對健康的影響，保證我們的健康。多食用抑制亞硝胺形成的食物，如吃香腸配大蒜，大蒜中的大蒜素可以抑制胃中的硝酸鹽還原菌，使胃內的亞硝酸鹽明顯降低；飲用如養樂多類的乳酸菌飲品，乳酸菌所產生的醋酸、乳酸能抑制硝酸鹽還原菌，阻斷致癌物質亞硝胺的形成，達到防癌的作用。茶葉中的茶多酚能夠阻斷亞硝胺的形成；富含維生素ｃ的食物，由於維生素C可防止胃中亞硝胺的形成，還有抑制亞硝胺的致突變作用。　　另外，吃烤香腸、臘肉、火腿、熱狗、培根等含亞硝酸鹽食物時，盡量不要與富含胺類的尤魚、秋刀魚、干貝、鱈魚等海鮮合吃，以避免在體內合成大量的亞硝胺，把握這個原則即可。--------------------------------------------------------------------------------　何謂『益生菌』( Probiotics )？　　　凡是能夠改善腸內菌叢的平衡，藉此而促進宿主健康作用的活性微　生物，皆稱為『益生菌』。　所有的乳酸菌都可以稱作為『益生菌』？　　　不是所有的乳酸菌都可稱為『益生菌』！必須符合及具備足夠條件　的菌株，才稱得上是『益生菌』喔！而『益生菌』要有效應用至少? 滬n　件標準：　　　原本就是人體腸內菌叢之一員。　　　能在腸中存活並繁殖。　　　能提供人體有益的效用。　　　產品製程及儲存期限可維持活性及存活率。　養樂多菌是符合及具備足夠條件的『益生菌』嗎？　　　是的，養樂多菌具有以下的特性：　　　原生性：原本就是人體腸內菌叢之一員。　　! 　耐酸性：能耐胃液、膽汁，於活的狀態到達腸內。　　　繁殖性：能在腸內繁殖。　　　效果性：對人體能發揮明顯有益的效果。　　　安全性：保證飲用的安全性及足夠的經驗。　　　安定性：製造成食品時亦能維持有效菌數。　　　簡易性：以便宜價格容易取得飲用。--------------------------------------------------------------------------------資料來源：養樂多企業網站http://www.yakult.com.tw",
        "title": "主題：吃香腸配養樂多會致癌嗎？ -  查無此人 - udn部落格",
        "topImageUrl": "https://g.udn.com.tw/community/img/common_all/blog_ogimg.jpg",
        "url": "http://blog.udn.com/watercmd/1066441",
      },
      Object {
        "canonical": "https://pension.president.gov.tw/cp.aspx?n=0710ED8C9356A871",
        "error": null,
        "summary": "網路流傳
			
			
			公教年金為何債台高築，據考試院不願具名官員私下表示，84年修法時巳刪除公教人員84年以後的年資退休後享有18%優惠存款的規定。卻在93年1月1日民進黨訂頒<政務人員退職撫卹條例>。就如蔡英文總統政大教授年資13年(73-79,82-89)，政務官5年(陸委會89-93,副院長95-97)，合計年資18年5個月，未達公教人員最低50歲，任職滿25年得退休的規定。由於民進黨政府93年訂頒<政務人員退職撫卹條例>, 蔡總統97年得以43歲副院長，年資18年5個月辦理退休，因退休時副院長薪資相當高，可辦18％優存金額410萬8600元全存入戶頭，每月的利息才62,829元。而93年訂頒的<政務人員退職撫卹條例>，規定可溯及既往，難怪當年起公保年金支出暴增，且持續增加，本來很多公務員要吃的這鍋公保年金就相當有限，突然跑來一堆大胃王（政務官）來搶食，公保年金不被吃跨才怪，由於這些人享有巨額退休金，就如馬前總統提到的，因而引起連戰和李元簇等不高興。

			民進黨的高官大口大口吞公保年金，如姚嘉文、游錫堃、吳榮義、郭瑤琪、杜正勝、陳哲男、朱武獻…..，口口聲聲罵18％，卻大手拿18％，真是無恥到極點。蘇貞昌也一再反18％，自稱沒有去領，問題錢入到他的戶頭就是他的，請問終止了該帳戶嗎？姚嘉文辯解說他雖享有18%優利存款，但他主張改革有什麼不對？公保年金被這些政務官大把大把搬光了，再來要求其他小公務員縮衣節食，硬要人多繳費少領錢，天底下就是有這麼不公平的事。
			
			
			這篇網路流傳的文章謬誤至為明顯如下：

			
				
				「政務人員退職酬勞金給與條例」於民國61年2月5日制定之初已有任滿2年可併計事務官年資領取一次退職金規定。至民國74年11月29日修正第四條又增列可領 月退職酬勞金規定。全文在民國88年6月15日修正，計算退職金基礎修正，且還回溯至民國85年5月1日開始適用。根本與93年施行的「政務人員退職撫卹條例」無關。
				
				
				而一次退職酬勞金存入18%優惠存款，更是民國61年10月26日就已訂定的「退職政務人員一次退職酬勞金優惠存款辦法」所規定。也不是93年施行的「政務人員退職撫卹條例」所訂定。政務官任滿2年、年資併計、退職酬勞金存18%都是在民國88年以前就存在的架構，已適用多年，且於93年後就已廢止18%優惠存款，改為一次離職金了。
				
				
				亦即，我國政務人員退職制度在92年以前，是採行「確定給付制」，自93年1月1日「政務人員退職撫卹條例」施行後，已改為「確定提撥之離職儲金制」。依現行規定，93年1月1日以後由軍、公、教人員、其他公職人員或公營事業人員轉任政務人員者，在職期間按月提撥離職儲金，退職時領取公、自提儲金本息（非由軍、公、教人員、其他公職人員或公營事業人員轉任政務人員者，無法參加政務人員離職儲金）；因此，93年以後擔任政務人員者，已無法領取退職酬勞金。目前只有少部分於92年以前已任政務人員且符合一定條件者，退職時仍可適用92年以前原規定請領退職酬勞金及辦理優惠存款。所以並無網路流傳文章所稱「政務人員退職撫卹條例規定讓政務人員享有巨額退休金」的情事。
				
				
				另外，就算是適用92年以前原規定請領退職酬勞金及辦理優惠存款的政務人員，其優惠存款金額也一體適用銓敘部於95年2月16日、100年1月1日及同年2月1日推動的調整方案進行調降，其中95年2月16日實施的方案對於支領月退職酬勞金的政務人員退休所得替代率上限訂為不超過現職待遇的75%至80%（較一般基層公務人員退休所得替代率上限為85%至95%更為嚴格），超過者必須調降養老給付優惠存款金額。至於100年2月1日實施的現行方案已明定支領月退職酬勞金的退職特任政務人員公保養老給付優惠存款金額不得超過200萬元。
				
			

			以上說明，自95年以來實施的歷次優惠存款調整方案，都一體適用於政務人員。這些因制度過渡期產生的人數不多，所占比率也小，不可能有如網路流傳文章所稱「公保年金被政務人員大把搬光，要求其他小公務員縮衣節食」的情事；更不應該成為年金制度不需改革的藉口。",
        "title": "輿情澄清4：政務人員退職金吃垮公保年金嗎?",
        "topImageUrl": "https://pension.president.gov.tw/images/share.png",
        "url": "https://pension.president.gov.tw/cp.aspx?n=0710ED8C9356A871",
      },
      Object {
        "canonical": "https://www.fun123.com.tw/site/promote/download_app",
        "error": null,
        "summary": "流行3C周邊商品 限時特價",
        "title": "3C市集 | 流行3C周邊商品 限時特價",
        "topImageUrl": "https://www.fun123.com.tw/images/desktop/downloadapp/fun123_downloadapp_banner.jpg",
        "url": "http://ms7.tw/DL/D?k=app_daily_coupon",
      },
      Object {
        "canonical": "http://blog.renren.com/share/481612507/14628038473",
        "error": null,
        "summary": "早泄能根治吗：　　生活中的许多苦难，让我们学会了承受，学会了担当，学会了在泪水中挺立自己的灵魂，在坚韧中亮化自己的人格。　　生活从来都是波澜起伏的，命运从来都是峰回路转的，因为有了曲折和故事，我们的生命才会精彩。　　有时候，哭泣，不是屈服；后退，不是认输；放手，不是放弃；沉默，不是无话可说。      早泄能根治吗-中国女性幸福指数调查报告      早泄能根治吗-ComeOn持久因子测试结果      早泄能根治吗-ComeOn持久因子效果咋样      我是一个内向的人，平时很少跟女孩子说话，有的时候说不了几句就会脸红，但是我的内心却很是渴望得到女孩子的关注和喜爱，这让我看起来非常的突兀和搞笑，因为这在大学同宿舍的人没少嘲笑我，确实我有的时候就是这样，自己还养成了手淫的习惯，对于手淫开始感觉挺爽的，但是越来越感到时间短，最后直接导致了早绁的发生，多亏有早绁喷剂ComeOn持久因子，早绁喷剂ComeOn持久因子让我不再犹豫不决。大学的时代，对于有的男孩子来说是一个乐园，但是对于我这样的男孩子来说，却是郁闷而烦恼比较多，所以一般的情况下，我都会是在宿舍和学校比较多一点，所以在外人看来我是一个老实话不多，而且非常腼腆的男孩子，当然的也没有女孩子会喜欢我，一说话就脸红的男孩子谁敢啊，所以我偷偷的手淫的时候宿舍的人都不清楚，还别说手淫的时候确实很爽，但是久而久之我发现，我的身体经常会出现腰疼的毛病，有的时候小解都成了问题，对于早绁，我一直都没有什么概念，因为根本没有跟人做过，但是我知道我可能已经有了早绁，对于早绁的治疗，我也试过很多方法，但是效果都不好，对于早绁的治疗，我也试过一些方法，但是效果都不是很好，再加上在宿舍本身就人多，我也不好意思，对于我的早绁问题一直搁置着。对于早绁的治疗，我也想试试一些比较新奇的方法，早绁喷剂ComeOn持久因子，我第一次听说的时候感到不是很靠谱，但是在观察了很久以后，我发现这个早绁喷剂ComeOn持久因子的口碑还是挺不错的呢，于是我就试着买了一盒早绁喷剂ComeOn持久因子，别说还真的挺好的，我用了一盒之后，去找那些学校外面的女人试试，还别说效果真的挺不错，于是我接着使用早绁喷剂ComeOn持久因子，别说用了一个疗程的早绁喷剂ComeOn持久因子以后，我身上的不适感完全没有了。对于早绁的治愈确实是一件高兴的事，但是令我更加兴奋的是我终于找到了自己心仪的人，我准备向她表明，虽然我不知道她是怎么想的，但是我有信心，早绁喷剂ComeOn持久因子让我不再犹豫不决 早泄能根治吗-ComeOn持久增强持久力      早泄能根治吗-ComeOn持久成分解析      早泄能根治吗-ComeOn持久只需三步      早泄能根治吗-ComeOn持久六大惊喜      早泄能根治吗-ComeOn持久七大保障      早泄能根治吗-【马上行动】      早泄能根治吗-【客户反馈】                早泄能根治吗-【现在就立刻行动】",
        "title": "（d7）早泄能根治吗？ComeOn持久因子金华哪里有卖【分享】 – 【人人分享-人人网】",
        "topImageUrl": "http://a.xnimg.cn/wap/mobile/2018activity/newr/xcx/bigxcx.jpg",
        "url": "http://blog.renren.com/share/300233063/14628038408",
      },
    ],
  },
}
`;
